pygmt/helpers/utils.py

```diff
@@ -187,23 +187,26 @@ def _check_encoding(
     return "ISOLatin1+"
 
 
-def data_kind(
+def data_kind(  # noqa: PLR0911
```

Comment:
PLR0911 means "too-many-return-statements".

https://docs.astral.sh/ruff/rules/too-many-return-statements/

Reply:
> This PR refactors the data_kind function, to use if-return statements instead of if-else statements. The new if-return statements are easier to read and maintain.

I don't see much of an improvement in readability/maintainability between the previous `if-elif-else` statements and the `if-return` style in this PR. Looking at https://stackoverflow.com/questions/72024229/how-to-solve-pylint-too-many-return-statements-elegantly, I'm thinking if we could either keep the `if-elif-else` syntax, or use `match-case` instead.

If using `match-case`, would it be more readable to do the `isinstance()` checks first, and then have a nested `if-elif` for `case _` for the more complex cases?

Reply:
I feel `match-case` would make it more complicated:
```
match data:
    case x if isinstance(x, io.StringIO):
        return "stringio"
    case x if isinstance(x, xr.DataArray):
        return "image" if len(x.dims) == 3 else "grid"
    case x if isinstance(x, bool | int | float) or (x is None and not required):
        return "arg"
```

Reply:
Using `match-case` also has the `too-many-return-statements` warning, and using `if-elif-else` means we need to type hints the `kind` variable like:
```
kind: Literal["arg", "file", "geojson", "grid", "image", "matrix", "stringio", "vectors"]
```

Reply:
Haven't run tests on this, but we can remove some of the `isinstance()` calls like so:

```python
match data:
        case str() | pathlib.PurePath():
            kind = "file"
        case list() | tuple() if all(
            isinstance(_file, str | pathlib.PurePath) for _file in data
        ):
            kind = "file"
        case io.StringIO():
            kind = "stringio"
        case (bool() | int() | float()) | None if not required:
            "arg"
        case xr.DataArray():
            kind = "image" if len(data.dims) == 3 else "grid"
        case x if hasattr(x, "__geo_interface__"):
            kind = "geojson"
        case x if x is not None:
            kind = "matrix"
        case _:
            kind = "vectors"
    return kind
```

See [class patterns](https://peps.python.org/pep-0634/#class-patterns) (xref https://stackoverflow.com/questions/67524641/convert-multiple-isinstance-checks-to-structural-pattern-matching/67524642#67524642)

Reply:
The `match-case` statements look more compact. Done in 33051ac88eb7e32119500ac8880f329d01be6d83.

The mypy warning is also suppressed.